workspaces/src/result.rs

```diff
@@ -100,15 +103,34 @@ impl CallExecutionDetails {
         }
     }
 
-    /// Convert the execution details into a Result if its status is not a successful one.
-    /// Useful for checking if the call was successful and forwarding the error upwards.
-    fn try_into_result(self) -> Result<Self> {
+    /// Checks whether the transaction was executed successfully. This will return
+    /// a `Ok(())` if the transaction was successful, and a `Err(Error)` if it was not.
+    ///
+    /// This acts a convenience function to allow us to forward errors when we do not
+    /// care about the details of the transaction at all.
+    pub fn executed(&self) -> Result<()> {
         self.try_into_success_value()?;
-        Ok(self)
+        Ok(())
     }
 
-    pub(crate) fn from_outcome(outcome: FinalExecutionOutcomeView) -> Result<Self> {
-        Self::from(outcome).try_into_result()
+    /// Converts the Execution details into a [`Result`]. [`Ok`] if the transaction succeed
+    /// and [`Err`] if it failed. The `Err` contains the [`Error`] along with the [`details`]
+    ///
+    /// [`details`]: crate::result::CallExecutionDetails
+    pub fn ok(self) -> Result<Self, (Self, Error)> {
+        match self.try_into_success_value() {
+            Ok(_) => Ok(self),
+            Err(err) => Err((self, err)),
+        }
+    }
+
+    /// Checks whether the transaction execution failed. This will return
+    /// `Some((CallExeuctionDetails, Error))` if it did fail, and `None` if it succeeded.
+    pub fn err(self) -> Option<(Self, Error)> {
+        match self.try_into_success_value() {
+            Err(err) => Some((self, err)),
+            Ok(_) => None,
+        }
     }
```

Comment:
I'm very skeptical about the naming and APIs for these. These do not seem intuitive to me.

So it seems like as a developer, you use `json`, `borsh`, or `raw_bytes` methods to get a result but `executed` if you don't care about the result data. Intuitively feels like this should be named more descriptively as `into_result` or something, where maybe we can return some of the data about the successful outcome that would otherwise be dropped.

As for the `ok` and `err` functions, I don't have any clear idea of how these would be used, and it's very confusing that they have different return types. This will likely be confusing with the [err](https://doc.rust-lang.org/std/result/enum.Result.html#method.err) and [ok](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) methods on the Rust `Result` type.

Reply:
definitely the naming could be improved. I feel like `executed` could stay since that conveys that we're expecting to have executed correctly and we include something like `into_result -> Result<SuccessfulExecutionOutcome>` if people want to make use of all the details. As for `ok`/`err`, not sure what's a better name. Maybe `handle`/`err` would be better where we can still keep `err` since that just returns Option like `std::result::Result` does

Reply:
Alright, instead of returning `(ExecutionDetails, Error)`, the details is now in the error object via `Error::detailed` and users would be able to grab it via `error.details()`. Got rid of `ok` as well since that didn't make sense and added `into_result() -> Result<ExecutionDetails>` instead. This should make for cleaner API. Let me know what you think